int.to_bytes, int.from_bytes handling of 'signed' arg

sebastien-riou opened the pull request:
those functions have three issues:
- `signed` is a positional argument, shall be a keyword arg
- `signed`=True is silently ignored
- int.to_bytes only: negative value with signed=False are silently accepted

this pull request fix those issues:
- `signed` is now a keyword argument
- `signed`=True throws `NotImplementedError`
- int.to_bytes reject negative value if `signed`=False

This is my first pull request in this project, please kindly let me know if something is not right.

I hope to make another pull request to actually implement `signed`=True.

sebastien-riou committed:
int.to_bytes, int.from_bytes handling of 'signed' arg

dlech commented on py/objint.c:
```diff
@@ -388,67 +388,93 @@ mp_obj_t mp_obj_int_binary_op_extra_cases(mp_binary_op_t op, mp_obj_t lhs_in, mp
-    if (!mp_obj_is_small_int(args[0])) {
-        mp_obj_int_to_bytes_impl(args[0], big_endian, len, data);
+    if (!mp_obj_is_small_int(pos_args[0])) {
+        assert(mp_obj_is_type(self_in, &mp_type_int));
```
should be `pos_args[0]` instead of `self_in`

dlech commented on py/objint.c:
```diff
@@ -388,67 +388,93 @@ mp_obj_t mp_obj_int_binary_op_extra_cases(mp_binary_op_t op, mp_obj_t lhs_in, mp
+    if (!mp_obj_is_small_int(pos_args[0])) {
+        assert(mp_obj_is_type(self_in, &mp_type_int));
+        mp_obj_int_t *val = MP_OBJ_TO_PTR(pos_args[0]);
+        if(val->mpz.neg){
```
space after `if` and before `{` (see failed code format test in github actions)

dlech commented on py/objint.c:
```diff
@@ -388,67 +388,93 @@ mp_obj_t mp_obj_int_binary_op_extra_cases(mp_binary_op_t op, mp_obj_t lhs_in, mp
+    bool big_endian = pos_args[2] != MP_OBJ_NEW_QSTR(MP_QSTR_little);
 
+    if (args[ARG_signed].u_bool) {
+        mp_raise_msg(&mp_type_NotImplementedError, MP_ERROR_TEXT("signed=True not implemented (to_bytes)"));
```
What would it take to implement this?

dlech commented on py/objint.c:
```diff
@@ -388,67 +388,93 @@ mp_obj_t mp_obj_int_binary_op_extra_cases(mp_binary_op_t op, mp_obj_t lhs_in, mp
     {
-        mp_int_t val = MP_OBJ_SMALL_INT_VALUE(args[0]);
+        mp_int_t val = MP_OBJ_SMALL_INT_VALUE(pos_args[0]);
+        if(val<0){
```
same here

dlech reviewed (commented):
The changes look good to me.

It would also be useful to add tests to check that we get the expected results.

stinos commented on py/objint.c:
```diff
@@ -388,67 +388,93 @@ mp_obj_t mp_obj_int_binary_op_extra_cases(mp_binary_op_t op, mp_obj_t lhs_in, mp
         delta = -1;
     }
-
+    if (args[ARG_signed].u_bool) {
```
Would be clearer to put this right after parsing the arguments, and more sensible as well: bit strange to first parse a bit, then have some other logic, then again return to inspecting arguments.

stinos commented on py/objint.c:
```diff
@@ -388,67 +388,93 @@ mp_obj_t mp_obj_int_binary_op_extra_cases(mp_binary_op_t op, mp_obj_t lhs_in, mp
     return mp_obj_new_str_from_vstr(&mp_type_bytes, &vstr);
 }
-STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(int_to_bytes_obj, 3, 4, int_to_bytes);
+STATIC MP_DEFINE_CONST_FUN_OBJ_KW(int_to_bytes_obj, 3, int_to_bytes);
```
This could break things for code which happened to use this, can't this stay as-is?

stinos commented:
Might be better to split this in 2 commits: `signed` argument handling, and the rejection of negative arguments. Pleas pay attention to the commit message format (see all other commits to get an idea of what it should be like).

stinos commented on py/objint.c:
```diff
@@ -388,67 +388,93 @@ mp_obj_t mp_obj_int_binary_op_extra_cases(mp_binary_op_t op, mp_obj_t lhs_in, mp
-    // TODO: Support signed param (assumes signed=False at the moment)
-    (void)n_args;
+STATIC mp_obj_t int_from_bytes(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    enum { ARG_signed };
```
Not 100% sure, but seeing this code is repeated below, it might save code size to put it in a small function which does the parsing and the check.

sebastien-riou commented:
thanks for your comments, sorry for the abusive copy and past.

I don't think it will take much more effort to implement this, I just wanted to have a simple pull request to see all the steps.

That said I think that throwing exception is much better than silently returning wrong results (I found out those errors while teaching my nephew how to use its micro:bit board... micropython ended up in education toolbox!)

I will add few more commit to correct all the things you identified

sebastien-riou committed:
py: int.to_bytes/from_bytes handling of 'signed' arg

Parsing code in common function.

sebastien-riou committed:
py: int.to_bytes support for MICROPY_LONGINT_IMPL_LONGLONG

sebastien-riou committed:
tests/basics: add test of int.to_bytes signed kwarg

sebastien-riou committed:
py: int.to_bytes fix for MICROPY_LONGINT_IMPL_LONGLONG

sebastien-riou commented:
@stinos 
@dlech 
4 tests are still failing but I don't see how to reproduce them locally nor how to fix them. In particular I am wondering about the following:
````
2 tests failed: uctypes_ptr_le uctypes_ptr_native_le
````
I could not find any occurence of "uctypes_ptr" in the whole "tests" folder. I have run ``make test_full`` and no test fails.

dlech commented:
Did you try building a 32-bit version of MicroPython (Unix port) on a 64-bit system?

Looking at the `uctypes_ptr_le` test, the line that fails is:

    buf = addr.to_bytes(uctypes.sizeof(desc), "little")

But `desc` describes a struct of 3 pointers so `sizeof(desc)` should be 3 * 4 (given 32-bit pointer size). Maybe this has uncovered a bug in uctypes?

sebastien-riou commented:
This is what is happening:
````
user@debian:~/micropython/ports/unix$ ./micropython ../../tests/extmod/uctypes_ptr_le.py
addr=-1213022192
Traceback (most recent call last):
  File "../../tests/extmod/uctypes_ptr_le.py", line 23, in <module>
OverflowError: can't convert negative int to unsigned (to_bytes)
````
previously the negative number was accepted even though signed=False and it was somehow working out.
So the correction would be to add signed=True but that's not implemented yet.
I will try to support signed=True as it looks more work to just live with the not implemented exception.

sebastien-riou committed:
py: int.to_bytes/from_bytes handling of 'signed' arg

Implementation of signed=True
Tests updated accordingly

sebastien-riou committed:
tests/extmod: fix uctypes_ptr_* tests

int.to_bytes now has strict range check, this was causing exceptions in those tests on 32 bit ports

sebastien-riou commented:
@dlech 
> Did you try building a 32-bit version of MicroPython (Unix port) on a 64-bit system?
> 
> Looking at the `uctypes_ptr_le.py` test, the line that fails is:
> 
> ```
> buf = addr.to_bytes(uctypes.sizeof(desc), "little")
> ```
> 
> But `desc` describes a struct of 3 pointers so `sizeof(desc)` should be 3 * 4 = 12 (given 32-bit pointer size). Maybe this has uncovered a bug in uctypes?

it seems `sizeof(desc)` is 4 on 32 bit system. the test now pass on 'coverage_32' but fails on 'nanbox'. 
I don't understand why 'nanbox' pass on my machine AND my 32bit VM and yet fails in CI.

dpgeorge commented on py/objint_mpz.c:
```diff
@@ -107,15 +107,77 @@ char *mp_obj_int_formatted_impl(char **buf, size_t *buf_size, size_t *fmt_size,
+        if (buf[msb] & 0x80) {
+            o->mpz.neg = 1;
+        }
+    }
```
I think this negation logic belongs in `mpz_set_from_bytes()`, it should be slightly smaller code there.

dpgeorge commented on py/objint_mpz.c:
```diff
@@ -107,15 +107,77 @@ char *mp_obj_int_formatted_impl(char **buf, size_t *buf_size, size_t *fmt_size,
+    if ( (len * 8) < bits_cnt ){
+        mp_raise_msg(&mp_type_OverflowError, MP_ERROR_TEXT("value too large (to_bytes)"));
+    }
     mpz_as_bytes(&self->mpz, big_endian, len, buf);
```
The `mpz_bit_length()` function is quite a lot of extra code.  I think `mpz_as_bytes()` can easily tell when the value will overflow, and can be changed to return a bool, whether it fit or not.

dpgeorge commented on py/makeqstrdata.py:
```diff
@@ -197,6 +197,7 @@
     "set",
     "setattr",
     "setdefault",
+    "signed",
```
Please don't add this, it requires changing the bytecode version number, and it not necessary to add this here.

dpgeorge commented:
Thanks for the contribution, and sorry this PR got neglected.

Overall, this feels like a lot of extra code to handle this signed argument (although see my comments above to help reduce code size).  And there has not been much interest in supporting this.  So we need to make a decision whether it's worth spending bytes of code size on implementing this.

dpgeorge commented:
Closing for now.  Feel free to reopen for further discussion.
